jena-arq/src/main/java/org/apache/jena/sparql/expr/nodevalue/NodeFunctions.java

```diff
@@ -139,31 +140,24 @@ public static NodeValue sameTerm(NodeValue nv1, NodeValue nv2) {
         return NodeValue.booleanReturn(sameTerm(nv1.asNode(), nv2.asNode())) ;
     }
 
-    public static boolean sameTerm(Node n1, Node n2) {
-        if ( n1.equals(n2) )
+    public static boolean sameTerm(Node node1, Node node2) {
+        if ( node1.equals(node2) )
             return true ;
-        if ( n1.isLiteral() && n2.isLiteral() ) {
-            // But language tags are case insensitive.
-            String lang1 = n1.getLiteralLanguage() ;
-            String lang2 = n2.getLiteralLanguage() ;
-
-            if ( !lang1.equals("") && lang1.equalsIgnoreCase(lang2) ) {
-                // Two language tags, equal by case insensitivity.
-                boolean b = n1.getLiteralLexicalForm().equals(n2.getLiteralLexicalForm()) ;
-                if ( b )
-                    return true ;
-            }
+        if ( Util.isLangString(node1) && Util.isLangString(node2) ) {
+            String lex1 = node1.getLiteralLexicalForm();
+            String lex2 = node2.getLiteralLexicalForm();
+            if ( !lex1.equals(lex2) )
+                return false;
+            return node1.getLiteralLanguage().equalsIgnoreCase(node2.getLiteralLanguage());
```

Comment:
I guess this is accounting for messy data with `"foo"@en-us` and `"foo"@en-US` since specs do say that language tags should be case insensitive

Reply:
Yes.

RDF actually says "The value space of language tags is always in lower case." which is a bit weasley. It also say "may convert to lower case".  RDF 1.0 says "the result of comparing two language tags should not be sensitive to the case of the original input.".

Jena in-memory, plain graphs do this as "value-matching" but that is for all values (e.g. integers). This is trying to be as compatible as possible without hard coding the rest of XSD value-matching.

Jena isn't perfect and consistent in this area and never has been. Short of a discontinuous change to `GraphMem`, this seems the best option.

The parsers have an option to lowercase language tags on input and also an option to convert to RFC 5646 canonical form ("en-US").